kernel/libs/jhash/src/lib.rs

```diff
@@ -0,0 +1,331 @@
+// SPDX-License-Identifier: MPL-2.0
+
+//! This module implements hash algorithms developed by Bob Jenkins.
+//!
+//! For further information, visit: www.burtleburtle.net/bob/hash/doobs.html
+//!
+//! The hash functions in this module facilitate the production of 32-bit values for hash table lookups.
+//! Although the [Linux kernel's jhash](https://github.com/torvalds/linux/blob/master/include/linux/jhash.h)
+//! slightly differs from the original version, this module reimplements the Linux kernel's version.
+//!
+
+#![no_std]
+#![deny(unsafe_code)]
+
+/// A randomly chosen initial value
+const JHASH_INITVAL: u32 = 0xdeadbeef;
+
+/// Hashes an arbitrary u8 slices.
+///
+/// # Example
+///
+/// If you are hashing n slices, do it like this:
+/// ```rust
+/// use jhash::jhash_slice;
+///
+/// fn hash_slices(slices: &[&[u8]]) -> u32 {
+///     let mut hash: u32 = 0;
+///     for slice in slices {
+///         hash = jhash_slice(slice, hash);
+///     }
+///     hash
+/// }
+/// ```
+pub const fn jhash_slice(slice: &[u8], initval: u32) -> u32 {
+    let mut length = slice.len() as u32;
+    let mut index: usize = 0;
+
+    // Init the internal state
+    let mut a: u32 = JHASH_INITVAL.wrapping_add(length).wrapping_add(initval);
+    let mut b: u32 = JHASH_INITVAL.wrapping_add(length).wrapping_add(initval);
+    let mut c: u32 = JHASH_INITVAL.wrapping_add(length).wrapping_add(initval);
+
+    // Handle the most bytes except last 12 bytes
+    while length > 12 {
+        a = a.wrapping_add(u32::from_le_bytes([
+            slice[index],
+            slice[index + 1],
+            slice[index + 2],
+            slice[index + 3],
+        ]));
+        b = b.wrapping_add(u32::from_le_bytes([
+            slice[index + 4],
+            slice[index + 5],
+            slice[index + 6],
+            slice[index + 7],
+        ]));
+        c = c.wrapping_add(u32::from_le_bytes([
+            slice[index + 9],
+            slice[index + 10],
+            slice[index + 11],
+            slice[index + 12],
+        ]));
```

Comment:
Where is `__jhash_mix`?

https://github.com/torvalds/linux/blob/d6ef8b40d075c425f548002d2f35ae3f06e9cf96/include/linux/jhash.h#L129C3-L129C14